src/StakeVault.sol

```diff
@@ -51,6 +52,20 @@ contract StakeVault is Ownable {
         stakeManager = _stakeManager;
     }
 
+    /**
+     * @notice Registers the vault with the stake manager.
+     */
+    function register() external {
+        stakeManager.registerVault();
+    }
```

Comment:
First I wanted to make `register()` part of the StakeVault's `constructor()` but the problem is `registerVault()` has the trusted code hash modifier. We need to create a `StakeVault` "template" instance first to register its codehash with the Stake manager, but the instantiation then fails due to `registerVault()` reverting.

I then decided to move this out and add an `onlyRegisteredVaults` modifier to the sensitive functions.

This should ensure `StakeVault` will call `register()` before they'll interact with the StakeManager.

Reply:
Since we only allow for staking once per vault, we could also make this part of `stake()` instead and enforce registration implicitly there.

Let me know what you guys think.

Reply:
> Since we only allow for staking once per vault, we could also make this part of stake() instead and enforce registration implicitly there.

I like this idea!

Otherwise we can do it in the constructor anyway since we know the bytecode already, we don't need to deploy a real one before. But anyway I like it if we do it in the `stake` function

Reply:
> Otherwise we can do it in the constructor anyway since we know the bytecode already, we don't need to deploy a real one before. But anyway I like it if we do it in the stake function

So I gave it a bit more thought.
When I started implementing it, it felt odd to me that `stake()` would create a side effect (registration).
Now I'm kind of hesitant..

re: the other point: we can't do it in the constructor because we don't know the deployed byte code at this point in time. This is because of the `STAKING_TOKEN` in stakevault being immutable, making it part of the deploycode.

We know the staking token is going to be SNT on mainnet/l2 but for local testing it's always a different one.

So I don't think we can make registration part of the constructor.

Reply:
> because we don't know the deployed byte code at this point in time

@0x-r4bbit we can add a script that compute it in local. It can use a mock manager without modifier so that the vault can be deployed in local and the bytecode can be printed out. after that we know it and we can use it in production. we don't need to deploy a contract in prod to know the bytecode, wdyt?

Reply:
> we don't need to deploy a contract in prod to know the bytecode, wdyt?

I think that would be idea.
If we do a minimal proxy clone pattern for `StakeVault` that needs a template deployed anyways, I think we don't need to set up a script for that either.

Maybe another option worth exploring?